File: src/lib/store/Store.java
package ddbt.lib.store;
import scala.Function1;
import scala.Unit;

/**
 * Index interface.
 * By default it emits a warning when an unsupported function is called.
 * Specific implementation are below.
 *
 * @author TCK
 */
abstract class Idx<E extends Entry> {
  Idx(int i,boolean u) { idx=i; unique=u; }
  protected final int idx;
  protected final boolean unique;
  protected int size;

  protected void w(String n) {}//println(this.getClass.getName+": "+n+" not supported")
  public void insert(E e) { w("insert"); }
  public void delete(E e) { w("delete"); }
  public void update(E e) { w("update"); } // reposition the entry if key/hash modified
  public E get(E key) { w("get"); return null; } // returns the first element only
  public void foreach(Function1<E,Unit> f) { w("foreach"); } // on all elements; warning: what about reordering updates?
  public void slice(E key,Function1<E,Unit> f) { w("slice"); } // foreach on a slice
  public void range(E min, E max, boolean withMin, boolean withMax, Function1<E,Unit> f) { w("range"); }
  public void clear() { w("clear"); }
  public void compact() { w("compact"); }
  public String info() { return "Idx("+idx+","+(unique?"unique":"multiple")+"){\n  type => ?\n}"; }
}

class IdxHashEntry<E extends Entry> {
  IdxHashEntry(int h, E d) { hash=h; data=d; }
  int hash; E data; IdxHashEntry<E> same, diff;
}

/**
 * IdxHash is a hashed index.
 * Buckets elements are stored in a special tree (invariant: sub-nodes of a same
 * branch do not have diff children). A bucket can be seen of a list(diff) of
 * list(same) of identical elements (a.cmp(idx,b)==0).
 * - If unique, a bucket is a simple list along 'diff'
 * - If non-unique, all elements along the 'same' branch are equal.
 *
 * Supported operations:
 * + insert,delete,update,get = O(1)
 * + range = O(n)
 */
@SuppressWarnings("unchecked")
class IdxHash<E extends Entry> extends Idx<E> {
  private static final int init_capacity = 16;
  private static final int max_capacity = 1 << 30;
  private static final float compact_factor = 0.05f;
  private final float load_factor;
  private IdxHashEntry<E>[] data = new IdxHashEntry[init_capacity];
  private int threshold;

  IdxHash(int i, boolean u) { super(i,u);
    load_factor = u ? 0.75f : 4.0f;
    threshold = Math.round(init_capacity * load_factor);
  }
  // Private/inlined functions
  // @inline private def _hash(e:E) = { var h=e.hash(idx); h^=(h>>>20)^(h>>>12)^(h<<9); h^(h>>>7)^(h>>>4); }
  private void _resize(int new_capacity) {
    IdxHashEntry<E>[] d=new IdxHashEntry[new_capacity];
    for (int i=0,n=data.length;i<n;++i) { IdxHashEntry<E> e=data[i], en;
      while (e!=null) { en=e.diff; int b=e.hash&(new_capacity-1); e.diff=d[b]; d[b]=e; e=en; }
    }
    data=d; threshold=Math.min(Math.round(new_capacity*load_factor), max_capacity+1);
  }
  private boolean _del(E e,IdxHashEntry<E> i) {
    IdxHashEntry<E> p=i.same; if (p!=null) { i.same=p.same; i.data=p.data; p.data.data[idx]=i; return true; } // eat same child
    p=i.diff; if (p!=null) { i.hash=p.hash; i.same=p.same; i.diff=p.diff; i.data=p.data; p.data.data[idx]=i; return true; } // eat diff child
    // delete from parent (i is a leaf)
    int h=i.hash, b=h&(data.length-1); p=data[b];
    if (i==p) { data[b]=null; return true; } // it's the root
    else do {
      if (i==p.diff) { p.diff=null; return true; } // leaf of diff branch
      else if (p.hash==h && e.cmp(idx,p.data)==0) do { IdxHashEntry<E> s=p.same; if (i==s) { p.same=null; return true; }; p=s; } while (p!=null); // leaf of same branch
      p=p.diff;
    } while(p!=null);
    return false;
  }
  // Public
  @Override public void insert(E e) {
    if (size==threshold) { int n=data.length; if (n==max_capacity) threshold=Integer.MAX_VALUE; else _resize(n << 1); }
    int h=e.hash(idx), b=h&(data.length-1);
    IdxHashEntry<E> p=data[b], i=new IdxHashEntry<E>(h,e); e.data[idx]=i;
    if (p==null) { data[b]=i; size+=1; return; }
    else do {
      if (p.hash==h && e.cmp(idx,p.data)==0) {
        if (unique) { p.data=e; e.data[idx]=p; }
        else { i.same=p.same; p.same=i; size+=1; return; }
        return;
      }
      if (p.diff==null) { p.diff=i; size+=1; return; }
      p=p.diff;
    } while(p!=null);
  }
  @Override public void delete(E e) { IdxHashEntry<E> i=(IdxHashEntry<E>)e.data[idx]; if (i!=null && _del(e,i)) size-=1; }
  @Override public void update(E e) { IdxHashEntry<E> i=(IdxHashEntry<E>)e.data[idx]; if (i!=null && i.hash!=e.hash(idx) && _del(e,i)) { size-=1; insert(e); } }
  @Override public E get(E key) { int h=key.hash(idx); IdxHashEntry<E> e=data[h&(data.length-1)];
    while (e!=null && (e.hash!=h || key.cmp(idx,e.data)!=0)) e=e.diff; return e!=null ? e.data : null;
  }
  @Override public void foreach(Function1<E,Unit> f) { E d; IdxHashEntry<E> e,en;
    for (int i=0,n=data.length;i<n;++i) { e=data[i];
      if (e!=null) do { en=e.diff; do { d=e.data; e=e.same; f.apply(d); } while(e!=null); e=en; } while(e!=null);
    }
  }
  @Override public void slice(E key,Function1<E,Unit> f) { int h=key.hash(idx); IdxHashEntry<E> e=data[h&(data.length-1)];
    while (e!=null && (e.hash!=h || key.cmp(idx,e.data)!=0)) e=e.diff;
    if (e!=null) do { E d=e.data; e=e.same; f.apply(d); } while (e!=null);
  }
  @Override public void range(E min, E max, boolean withMin, boolean withMax, Function1<E,Unit> f) {
    int cMin=withMin?-1:0; int cMax=withMax?1:0;
    for (int i=0,n=data.length;i<n;++i) { IdxHashEntry<E> e=data[i],em,en;
      while (e!=null) { en=e.diff; if (e.data.cmp(idx,min)>cMin && e.data.cmp(idx,max)<cMax) do { em=e.same; f.apply(e.data); e=em; } while (e!=null); e=en; }
    }
  }
  @Override public void clear() { IdxHashEntry<E> z=null; for(int i=0,n=data.length;i<n;++i) { data[i]=z; } size=0; }
  @Override public void compact() { if (data.length*compact_factor>size) _resize(Math.max(init_capacity, 1 << ((int)Math.ceil(1+(Math.log((size/load_factor))/Math.log(2)))) )); }
  @Override public String info() { IdxHashEntry<E> e,en; int max=0; int n=data.length;
    for (int i=0;i<n;++i) { e=data[i]; int c=0; if (e!=null) do { en=e.diff; do { ++c; e=e.same; } while(e!=null); e=en; } while(e!=null); if (c>max) max=c; }
    return "IdxHash("+idx+","+(unique?"unique":"multiple")+") {\n  elements => "+size+"\n  buckets => "+n+"\n  occupancy => "+(size*1.0/n)+"\n  occupancyMax => "+max+"\n  loadFactor => "+load_factor+"\n}";
  }
}

File: src/lib/spark/store/Store.scala
package ddbt.lib.spark.store

import scala.annotation.elidable
import scala.annotation.elidable.ASSERTION
import scala.collection.mutable.HashMap
import scala.reflect.ClassTag

/**
 * A store contrains entries and allow access through multiple indices.
 * + The entries need extends Entry and specialize to define its fields and
 *   projections that will be used for indexing.
 * + A store operates over entries and can only retrieve data through indices,
 *   hence at least one index must exist.
 * + Indices can be modified from one implementation to another to dynamically
 *   support different workloads during the application lifecycle.
 * Note: to enforce proper inlining, compile with the '-optimize' option.
 *
 * @author TCK
 */

// XXX: throw an exception if update/delete/insert during foreach ?
// XXX: do we want/require the fallback to packed mode for IDirect ?
// XXX: ignore strict comparison [cmp] to use only hash comparison in IDirect?

/** Index types that can be defined over the data */
abstract sealed class IndexType
case object INone   extends IndexType // no index
case object IHash   extends IndexType // hash index, does not support range/min/max/median operations


object Store {
  val GATHER_STATISTICS=false
  // DBToaster: create n+1 hash indexes (n=# projections)
  //def apply[E<:Entry](n:Int,ops:EntryOps[E]=null)(implicit cE:ClassTag[E]) = new Store((0 to n).map(i=>new IdxHash[E](ops,i,i==0)).toArray.asInstanceOf[Array[Idx[E]]])

  def ind(s:String,n:Int=1) = { val i="  "*n; i+s.replaceAll("\n? *$","").replaceAll("\n","\n"+i) }
  
  val totalTimersForStores:HashMap[String,HashMap[String,(Long,Int)]] = new HashMap[String,HashMap[String,(Long,Int)]]()
  val timersPerIndexForStores:HashMap[String,HashMap[String,HashMap[Int,(Long,Int)]]] = new HashMap[String,HashMap[String,HashMap[Int,(Long,Int)]]]()
  def printTimersInfo = {
    val res = new StringBuilder
    totalTimersForStores.foreach { case (store,totalTimer) =>
      res.append(store).append(": {\n")
      totalTimer.foreach{ case (op, (time,count)) =>
        res.append("  ").append(op).append(" (total time, # of calls, average time): (%d,%d,%d)".format(time,count,(time.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int])).append("\n")
      }
      timersPerIndexForStores.get(store) match {
        case Some(timersPerIndex) => timersPerIndex.foreach { case (op, indexTimers) =>
          indexTimers.foreach { case (idx, (time,count)) =>
            res.append("  ").append(op).append("(%d) (total time, # of calls, average time): (%d,%d,%d)".format(idx,time,count,(time.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int])).append("\n")
          }
        }
        case None => //don't care
      }
      res.append("}\n")
    }
    timersPerIndexForStores.foreach { case (store, timersPerIndex) =>
      totalTimersForStores.get(store) match {
        case Some(_) => //don't care
        case None => {
          res.append(store).append(": {\n")
          timersPerIndex.foreach { case (op, indexTimers) =>
            indexTimers.foreach { case (idx, (time,count)) =>
              res.append("  ").append(op).append("(%d) (total time, # of calls, average time): (%d,%d,%d)".format(idx,time,count,(time.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int])).append("\n")
            }
          }
          res.append("}\n")
        }
      }
    }
    ind(res.toString)
  }
  def printTimersInfoCSV = {
    val res = new StringBuilder
    totalTimersForStores.foreach { case (store,totalTimer) =>
      totalTimer.foreach{ case (op, (time,count)) =>
        res.append(store).append(",").append(op).append(",-1,%d,%d,%d".format(time,count,(time.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int])).append("\n")
      }
      timersPerIndexForStores.get(store) match {
        case Some(timersPerIndex) => timersPerIndex.foreach { case (op, indexTimers) =>
          indexTimers.foreach { case (idx, (time,count)) =>
            res.append(store).append(",").append(op).append(",%d,%d,%d,%d".format(idx,time,count,(time.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int])).append("\n")
          }
        }
        case None => //don't care
      }
    }
    timersPerIndexForStores.foreach { case (store, timersPerIndex) =>
      totalTimersForStores.get(store) match {
        case Some(_) => //don't care
        case None => {
          timersPerIndex.foreach { case (op, indexTimers) =>
            indexTimers.foreach { case (idx, (time,count)) =>
              res.append(store).append(",").append(op).append(",%d,%d,%d,%d".format(idx,time,count,(time.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int])).append("\n")
            }
          }
        }
      }
    }
    ind(res.toString)+"\n"
  }
  def addTimersFromStore(storeName:String, totalTimer:HashMap[String,(Long,Int)], timersPerIndex:HashMap[String,HashMap[Int,(Long,Int)]]) = {
    totalTimersForStores.get(storeName) match {
      case Some(currentTotalTimer) => {
        currentTotalTimer.foreach{ case (op, (currentTime,currentCount)) =>
          totalTimer.get(op) match {
            case Some((time,count)) => currentTotalTimer.update(op, (time+currentTime,count+currentCount))
            case None => //don't care
          }
        }
        totalTimer.foreach{ case (op, timeCount) =>
          currentTotalTimer.get(op) match {
            case Some(_) => //don't care
            case None => currentTotalTimer += (op -> timeCount)
          }
        }
      }
      case None => totalTimersForStores += (storeName -> totalTimer)
    }

    timersPerIndexForStores.get(storeName) match {
      case Some(currentTimersPerIndex) => {
        currentTimersPerIndex.foreach{ case (op, currentIndexTimers) =>
          timersPerIndex.get(op) match {
            case Some(indexTimers) => {
              currentIndexTimers.foreach{ case (idx, (currentTime,currentCount)) =>
                indexTimers.get(idx) match {
                  case Some((time,count)) => currentIndexTimers.update(idx, (time+currentTime,count+currentCount))
                  case None => //don't care
                }
              }
              indexTimers.foreach{ case (idx, timeCount) =>
                currentIndexTimers.get(idx) match {
                  case Some(_) => //don't care
                  case None => currentIndexTimers += (idx -> timeCount)
                }
              }
            }
            case None => //don't care
          }
        }

        timersPerIndex.foreach{ case (op, indexTimers) =>
          currentTimersPerIndex.get(op) match {
            case Some(_) => //don't care
            case None => {
              val currentIndexTimers = new HashMap[Int,(Long,Int)]
              indexTimers.foreach{ case (idx, timeCount) =>
                currentIndexTimers += (idx -> timeCount)
              }
              currentTimersPerIndex += (op -> currentIndexTimers)
            }
          }
        }
      }
      case None => timersPerIndexForStores += (storeName -> timersPerIndex)
    }
  }
}

/**
 * The store is the main structure to store entries. It requires at least one
 * index. Index usually retain all the data according to their specialization.
 * Index 0 should retain all data (and possibly be one of the fastest index).
 * Indices data structure can be dynamically adapted using the index() function.
 */
class Store[E<:MapEntry](val indices: Vector[Index[E]])(implicit cE:ClassTag[E]) {
  assert(indices.size > 0)

  @inline private def time[R](f:String)(block: => R):R = { block }
  @inline private def time[R](f:String, idx:Int)(block: => R):R = { block }

  // Compatibility with the old version
  val idxs = indices

  private val numIndices = indices.length
  
  def unsafeInsert(idx:Int,e:E):Unit = time("unsafeInsert") { if (e==null) return; indices(idx).unsafeInsert(e); var i=0; while(i < numIndices) { if (idx!=i && indices(i)!=null) indices(i).insert(e); i+=1; } }
  def insert(e:E):Unit = time("insert") { if (e==null) return; var i=0; while(i < numIndices) { if (indices(i)!=null) indices(i).insert(e); i+=1; } }
  def update(e:E):Unit = time("update") { if (e==null) return; var i=0; while(i < numIndices) { if (indices(i)!=null) indices(i).update(e); i+=1; } } // e already in the Store, update in foreach is _NOT_ supported
  def delete(e:E):Unit = time("delete") { if (e==null) return; var i=0; while(i < numIndices) { if (indices(i)!=null) indices(i).delete(e); i+=1; } } // e already in the Store
  def get(idx:Int,key:E):E = time("get",idx) { if (key==null) return key; indices(idx).get(key) }
  def foreach(f:E=>Unit):Unit = time("foreach") { indices(0).foreach(f) } // assumes idxs(0) is the most efficient index
  def slice(idx:Int,key:E,f:E=>Unit) = time("slice",idx) { if (key!=null) indices(idx).slice(key,f) }
  def range(idx:Int,min:E,max:E,withMin:Boolean=true,withMax:Boolean=true,f:E=>Unit) = time("range", idx) { indices(idx).range(min,max,withMin,withMax,f) }
  def delete(idx:Int,key:E):Unit = time("delete", idx) { slice(idx,key,e=>delete(e)) }
  def clear = time("clear") { var i=0; while(i < numIndices) { if (indices(i)!=null) indices(i).clear; i+=1; } }
  def compact = time("compact") { var i=0; while(i < numIndices) { if (indices(i)!=null) indices(i).compact; i+=1; } }
  def size = indices(0).size
  def ++(other: Store[E]) = time("concatenate") { other.foreach (insert); this }

  def getInfoStr:String = {
    val res = new StringBuilder("MapInfo => {\n")
    res.append("  size => ").append(if(indices(0) == null) indices(1).size else size).append("\n")
    res.append("  idxs => [\n")
    // totalTimers.foreach{ case (f,(t,count)) =>
    //   val avg = (t.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int]
    //   res.append("    time in ").append(f).append(" => (").append(t/1000000).append(".").append((t/1000)%1000).append(" ms spent for ").append(count).append(" calls) in average -> ").append(avg).append(" ns per call").append("\n")
    // }
    indices.zipWithIndex.foreach { case (idx, idxID) =>
      if(idx == null) res.append(Store.ind(Store.ind("INone"))).append(",\n")
      else {
        res.append(Store.ind(Store.ind(idx.info)))
        // res.append(" --> {\n")
        // timersPerIndex.foreach{ case (f,fMap) =>
        //   if(fMap.contains(idxID)) {
        //     val (t,count) = fMap(idxID)
        //     val avg = (t.asInstanceOf[Double] / count.asInstanceOf[Double]).asInstanceOf[Int]
        //     res.append("      time in ").append(f).append(" => (").append(t/1000000).append(".").append((t/1000)%1000).append(" ms spent for ").append(count).append(" calls) in average -> ").append(avg).append(" ns per call").append("\n")
        //   }
        // }
        // res.append("    }\n")
      }
    }
    res.append("]")
    res.append("}").toString
  }
}

// ------------------------------------------
// Single hash-index entries
abstract class Entry1[E<:Entry1[E]] {
  val hash:Int // hash function
  var next:E = null.asInstanceOf[E]
  def merge(e:E):Boolean // combine in this the value of e if keys are equal
}

class Store1[E<:Entry1[E]]()(implicit cE:ClassTag[E]) {
  private final val init_capacity = 16
  private final val max_capacity = 1 << 30
  private final val load_factor = 0.75f
  private var size = 0
  private var data = new Array[E](init_capacity)
  private var threshold = (init_capacity * load_factor).toInt
  def add(e:E) {
    // resize array if needed
    if (size==threshold) {
      val n=data.size; if (n==max_capacity) threshold=java.lang.Integer.MAX_VALUE
      else { val new_capacity = n << 1; val d=new Array[E](new_capacity)
        var i=0; while(i < n) { var e=data(i); while (e!=null) { val ne=e.next; val b=e.hash&(new_capacity-1); e.next=d(b); d(b)=e; e=ne }; i+=1 }
        data=d; threshold=Math.min((new_capacity*load_factor).toInt, max_capacity+1);
      }
    }
    // insert and merge equivalent entries
    val b=e.hash&(data.length-1);
    var p=data(b);
    if (p==null) { data(b)=e; size+=1; }
    else do {
      if (p.merge(e)) return
      var n=p.next; if (n==null) { p.next=e; size+=1; return; }; p=n;
    } while (p!=null)
  }
  def foreach(f:E=>Unit) { val n=data.length; var i=0; while(i < n) { var e=data(i); while (e!=null) { f(e); e=e.next }; i+=1 } }
  def clear() { data=new Array[E](init_capacity); size=0; threshold=(init_capacity * load_factor).toInt }
}
